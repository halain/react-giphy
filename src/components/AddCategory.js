import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ( {setCategories} ) => {
    
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) =>{
        setInputValue( e.target.value );
    }

    const handleSubmit = (e) =>{
        e.preventDefault();
        //console.log('submit');

        if (inputValue.trim().length>=2){
            //setCategories viene desde el componente padre. Se agrega el valor del input a las categorias mediante el operador ...
            setCategories( cat=>{
               return [inputValue, ...cat];
            });
            setInputValue('');
        }

    }
    
    return (
        <form onSubmit={handleSubmit}>
          <input 
                type="text"
                value={inputValue}  
                onChange={ handleInputChange }
           />  
        </form>
    )
}

AddCategory.protoTypes = {
    setCategories: PropTypes.func.isRequired
}