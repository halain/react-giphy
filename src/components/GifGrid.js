//import React, { useState, useEffect } from 'react';
import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';
//import {getGifs} from '../helpers/getGifs'

export const GifGrid = ({category}) => {

   // const [images, setImages] = useState([]);
   

    //se dispara una unica vez cuando se renderiza el componente, asi si hay cambios en el componente no ejecuta constantemente la funcion con cada cambio
    //** */ useEffect( () =>{ funcion_a_ejecutar },[lista de dependencias]), si cambia la dependecia (category) se ejecuta nuevamente la funcion 
//     useEffect( () =>{
//         getGifs(category)
//             .then(setImages);
//    //         .then(imgs => setImages(imgs));
//     },[ category ]) 

   


    //** */ Crear custom Hook. Permitira que cuando el comp cargue, realizar la peticion fetch e indique cuando esta cargando y cuando termino
    //const state = useFetchGifs();
    //console.log(state);
    const {data:images, loading} = useFetchGifs(category);
    console.log(images, loading);

    return (
      <>
        <h3 className="animate__animated animate__fadeIn">{category}</h3>

        {loading && <p className="animate__animated animate__flash">Loading...</p>}

        <div className="card-grid">
          {images.map((img) => (
            <GifGridItem key={img.id} {...img} /> //...img envia cada propiedad de img en un objeto nuevo al componente GifGridItem
          ))}
        </div>

        
        
      </>
    );
}
