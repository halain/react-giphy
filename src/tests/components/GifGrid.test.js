import { shallow } from 'enzyme';

import '@testing-library/jest-dom';
import React from 'react';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs'); //mock data , finje llamada a useFetchGifs


describe('pruebas en <GifGrid />', () => {

    const category = 'Hard';

    test('debe renderizarce correctamente el componente ', () => {

        useFetchGifs.mockReturnValue({
            data:[],
            loading: true
        });

        const wrapper = shallow(<GifGrid category = {category}  />);

        expect( wrapper ).toMatchSnapshot();

    });


    test('debe mostrar los items cuando se carguen las imagenes con useFetchGifs', () => {

        const imgs = [{
            id: 'ABC',
            url: 'http://loquesea.asca.co',
            title: 'Titulo prueba'
        },
        {
            id: '123',
            url: 'http://loquesea.123.co',
            title: 'Titulo prueba otro'
        }]

        useFetchGifs.mockReturnValue({
            data: imgs,
            loading: false
        });
       
        const wrapper = shallow(<GifGrid category = {category}  />);

        expect(wrapper).toMatchSnapshot(); 

        expect( wrapper.find('p').exists() ).toBe( false ); //el p del loading no debe exitir xk ya hay data

        expect( wrapper.find('GifGridItem').length).toBe(imgs.length);

    });
    
    
    
})
