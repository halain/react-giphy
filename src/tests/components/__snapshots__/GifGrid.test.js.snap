// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <GifGrid /> debe mostrar los items cuando se carguen las imagenes con useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Hard
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Titulo prueba"
      url="http://loquesea.asca.co"
    />
    <GifGridItem
      id="123"
      key="123"
      title="Titulo prueba otro"
      url="http://loquesea.123.co"
    />
  </div>
</Fragment>
`;

exports[`pruebas en <GifGrid /> debe renderizarce correctamente el componente  1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Hard
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
